src/libcore/str/pattern.rs

```diff
@@ -630,3 +509,430 @@ impl<'a, F> Pattern<'a> for F where F: FnMut(char) -> bool {
+                }
+            }
+
+            // We have found a match!
+            let match_pos = self.position;
+            self.position += needle.len(); // add self.period for all matches
+            if !long_period {
+                self.memory = 0; // set to needle.len() - self.period for all matches
+            }
+
+            return SearchStep::Match(match_pos, match_pos + needle.len());
+        }
+        SearchStep::Reject(old_pos, self.position)
+    }
+
+    // Follows the ideas in `next()`.
+    //
+    // All the definitions are completely symmetrical, with period(x) = period(reverse(x))
+    // and local_period(u, v) = local_period(reverse(v), reverse(u)), so if (u, v)
+    // is a critical factorization, so is (reverse(v), reverse(u)). Similarly,
+    // the "period" stored in self.period is the real period if long_period is
+    // false, and so is still valid for a reversed needle, and if long_period is
+    // true, all the algorithm requires is that self.period is less than or
+    // equal to the real period, which must be true for the forward case anyway.
+    //
+    // To search in reverse through the haystack, we search forward through
+    // a reversed haystack with a reversed needle, and the above paragraph shows
+    // that the precomputed parameters can be left alone.
+    #[inline]
+    fn next_back(&mut self, haystack: &[u8], needle: &[u8]) -> SearchStep {
+        // `next_back()` uses `self.end` as its cursor -- so that `next()` and `next_back()`
+        // are independent.
+        let old_end = self.end;
+        if old_end == 0 {
+            return SearchStep::Done
+        }
+        'search: loop {
+            // Check that we have room to search in
+            if needle.len() > self.end {
+                self.end = 0;
+                break 'search;
+            }
+
+            // Quickly skip by large portions unrelated to our substring
+            if !self.byteset_contains(haystack[self.end - needle.len()]) {
+                self.end -= needle.len();
+                continue 'search;
+            }
+
+            if old_end != self.end {
+                break 'search;
+            }
+
+            // See if the left part of the needle matches
+            for i in (0..self.crit_pos).rev() {
+                if needle[i] != haystack[self.end - needle.len() + i] {
+                    self.end -= self.crit_pos - i;
+                    continue 'search;
+                }
+            }
+
+            // See if the right part of the needle matches
+            for i in self.crit_pos..needle.len() {
+                if needle[i] != haystack[self.end - needle.len() + i] {
+                    self.end -= self.period;
+                    continue 'search;
+                }
+            }
+
+            // We have found a match!
+            let match_pos = self.end - needle.len();
+            self.end -= needle.len(); // add self.period for all matches
```

Comment:
To update this comment for `next_back` and make it more clear, maybe this should read "subtract self.period instead of needle.len() for overlapping matches" (similarly, the `next` case could be changed, but with add, not subtract)